Add support for xbstream usage

https://www.percona.com/doc/percona-xtrabackup/2.4/xbstream/xbstream.html
https://www.percona.com/doc/percona-xtrabackup/2.4/xbstream/xbstream.html
https://www.percona.com/doc/percona-xtrabackup/2.4/innobackupex/streaming_backups_innobackupex.html

We should find a way to implement this functionality. It will ease to test and use it directly.

The result of decrypting using xbstream:

```
2017-03-29 08:22:48 DEBUG    Installed Server is MySQL, will continue as usual.
2017-03-29 08:22:48 DEBUG    Using xbstream to extract and decrypt from full_backup.stream!
2017-03-29 08:22:48 DEBUG    The following xbstream command will be executed /home/shahriyar.rzaev/Percona_Xtrabackups/xb_2.4/usr/local/xtrabackup/bin/xbstream -x --parallel=100 --decrypt=AES256 --encrypt-key='VVTBwgM4UhwkTTV98fhuj+D1zyWoA89K' --encrypt-threads='VVTBwgM4UhwkTTV98fhuj+D1zyWoA89K'  --encrypt-chunk-size=65536< /home/shahriyar.rzaev/backup_dir/full/2017-03-29_08-22-13/full_backup.stream -C /home/shahriyar.rzaev/backup_dir/full/2017-03-29_08-22-13
2017-03-29 08:22:48 ERROR    XBSTREAM COMMAND FAILED!
2017-03-29 08:22:53 ERROR    xbstream: [ERROR] Unknown suffix 'V' used for variable 'encrypt-threads' (value 'VVTBwgM4UhwkTTV98fhuj+D1zyWoA89K')
xbstream: [Warning] option 'encrypt-threads': signed value 0 adjusted to 1
xbstream: [ERROR] /home/shahriyar.rzaev/Percona_Xtrabackups/xb_2.4/usr/local/xtrabackup/bin/xbstream: Error while setting value 'VVTBwgM4UhwkTTV98fhuj+D1zyWoA89K' to 'encrypt-threads'
```

Seems to be Python wrong variable assignment.

--- a/autoxtrabackup.py
+++ b/autoxtrabackup.py
@@ -9,9 +9,9 @@ import time
 import re
 import os
 import humanfriendly
-
 import logging
 import logging.handlers
+
 
 logger = logging.getLogger('')
 

--- a/backup_prepare/prepare.py
+++ b/backup_prepare/prepare.py
@@ -10,6 +10,7 @@ import shutil
 import time
 from general_conf.generalops import GeneralClass
 from general_conf.check_env import CheckEnv
+from os.path import isfile
 
 import logging
 logger = logging.getLogger(__name__)
@@ -342,6 +343,34 @@ class Prepare(GeneralClass):
                         logger.debug(
                             "####################################################################################################")
                         time.sleep(3)
+
+                    # Extracting streamed incremental backup prior to executing new incremental backup
+
+                    if hasattr(self, 'stream') and isfile(
+                                    ("%s/%s/full_backup.stream") % (self.inc_dir, i)):
+                        logger.debug("Using xbstream to extract from full_backup.stream!")
+                        xbstream_command = "%s %s < %s/%s/inc_backup.stream -C %s/%s" % (
+                            self.xbstream,
+                            self.xbstream_options,
+                            self.inc_dir,
+                            i,
+                            self.inc_dir,
+                            i
+                        )
+
+                        logger.debug(
+                            "The following xbstream command will be executed %s",
+                            xbstream_command)
+
+                        status, output = subprocess.getstatusoutput(xbstream_command)
+                        if status == 0:
+                            logger.debug("XBSTREAM command succeeded.")
+                        else:
+                            logger.error("XBSTREAM COMMAND FAILED!")
+                            time.sleep(5)
+                            logger.error(output)
+                            return False
+
 
                         # Check if decryption enabled
                         if hasattr(self, 'decrypt'):

--- a/backup_prepare/prepare.py
+++ b/backup_prepare/prepare.py
@@ -344,32 +344,32 @@ class Prepare(GeneralClass):
                             "####################################################################################################")
                         time.sleep(3)
 
-                    # Extracting streamed incremental backup prior to executing new incremental backup
+                        # Extracting streamed incremental backup prior to executing new incremental backup
 
-                    if hasattr(self, 'stream') and isfile(
-                                    ("%s/%s/full_backup.stream") % (self.inc_dir, i)):
-                        logger.debug("Using xbstream to extract from full_backup.stream!")
-                        xbstream_command = "%s %s < %s/%s/inc_backup.stream -C %s/%s" % (
-                            self.xbstream,
-                            self.xbstream_options,
-                            self.inc_dir,
-                            i,
-                            self.inc_dir,
-                            i
-                        )
+                        if hasattr(self, 'stream') and isfile(
+                                        ("%s/%s/inc_backup.stream") % (self.inc_dir, i)):
+                            logger.debug("Using xbstream to extract from full_backup.stream!")
+                            xbstream_command = "%s %s < %s/%s/inc_backup.stream -C %s/%s" % (
+                                self.xbstream,
+                                self.xbstream_options,
+                                self.inc_dir,
+                                i,
+                                self.inc_dir,
+                                i
+                            )
 
-                        logger.debug(
-                            "The following xbstream command will be executed %s",
-                            xbstream_command)
+                            logger.debug(
+                                "The following xbstream command will be executed %s",
+                                xbstream_command)
 
-                        status, output = subprocess.getstatusoutput(xbstream_command)
-                        if status == 0:
-                            logger.debug("XBSTREAM command succeeded.")
-                        else:
-                            logger.error("XBSTREAM COMMAND FAILED!")
-                            time.sleep(5)
-                            logger.error(output)
-                            return False
+                            status, output = subprocess.getstatusoutput(xbstream_command)
+                            if status == 0:
+                                logger.debug("XBSTREAM command succeeded.")
+                            else:
+                                logger.error("XBSTREAM COMMAND FAILED!")
+                                time.sleep(5)
+                                logger.error(output)
+                                return False
 
 
                         # Check if decryption enabled

--- a/backup_prepare/prepare.py
+++ b/backup_prepare/prepare.py
@@ -60,6 +60,65 @@ class Prepare(GeneralClass):
             logger.debug(
                 "################################################################################################")
 
+            # Extract and decrypt streamed full backup prior to executing incremental backup
+            if hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") %
+                                                          (self.full_dir, self.recent_full_backup_file())) \
+                    and hasattr(self, 'encrypt') \
+                    and hasattr(self, 'xbs_decrypt'):
+                logger.debug("Using xbstream to extract and decrypt from full_backup.stream!")
+                xbstream_command = "%s %s --decrypt=%s --encrypt-key=%s --encrypt-threads=%s " \
+                                   "< %s/%s/full_backup.stream -C %s/%s" % (
+                                       self.xbstream,
+                                       self.xbstream_options,
+                                       self.decrypt,
+                                       self.encrypt_key,
+                                       self.encrypt_threads,
+                                       self.full_dir,
+                                       self.recent_full_backup_file(),
+                                       self.full_dir,
+                                       self.recent_full_backup_file()
+                                   )
+
+                logger.debug(
+                    "The following xbstream command will be executed %s",
+                    xbstream_command)
+
+                status, output = subprocess.getstatusoutput(xbstream_command)
+                if status == 0:
+                    logger.debug("XBSTREAM command succeeded.")
+                else:
+                    logger.error("XBSTREAM COMMAND FAILED!")
+                    time.sleep(5)
+                    logger.error(output)
+                    return False
+
+            # Extract streamed full backup prior to executing incremental backup
+            elif hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") %
+                                                            (self.full_dir, self.recent_full_backup_file())):
+                logger.debug("Using xbstream to extract from full_backup.stream!")
+                xbstream_command = "%s %s < %s/%s/full_backup.stream -C %s/%s" % (
+                    self.xbstream,
+                    self.xbstream_options,
+                    self.full_dir,
+                    self.recent_full_backup_file(),
+                    self.full_dir,
+                    self.recent_full_backup_file()
+                )
+
+                logger.debug(
+                    "The following xbstream command will be executed %s",
+                    xbstream_command)
+
+                status, output = subprocess.getstatusoutput(xbstream_command)
+                if status == 0:
+                    logger.debug("XBSTREAM command succeeded.")
+                else:
+                    logger.error("XBSTREAM COMMAND FAILED!")
+                    time.sleep(5)
+                    logger.error(output)
+                    return False
+
+
             # Check if decryption enabled
             if hasattr(self, 'decrypt'):
                 if hasattr(self, 'remove_original_enc') and (self.remove_original_enc):
@@ -344,7 +403,7 @@ class Prepare(GeneralClass):
                             "####################################################################################################")
                         time.sleep(3)
 
-                        # Extracting streamed incremental backup prior to executing new incremental backup
+                        # Extracting streamed incremental backup prior to preparing
 
                         if hasattr(self, 'stream') and isfile(
                                         ("%s/%s/inc_backup.stream") % (self.inc_dir, i)):

--- a/general_conf/bck.conf
+++ b/general_conf/bck.conf
@@ -54,6 +54,13 @@ xtra_prepare=--apply-log-only
 #Enable if you want to remove .qp files after decompression.(Not available yet, will be released with XB 2.3.7 and 2.4.6)
 #remove_original=FALSE
 
+[Xbstream]
+# Enable this, if you want to stream your backups
+#stream=xbstream
+#parallel=16
+# WARN, enable this, if you want to stream your backups to remote host
+#remote_stream=ssh xxx.xxx.xxx.xxx
+
 #Optional remote syncing
 #[Remote]
 #remote_conn=root@xxx.xxx.xxx.xxx

--- a/general_conf/bck.conf
+++ b/general_conf/bck.conf
@@ -55,10 +55,12 @@ xtra_prepare=--apply-log-only
 #remove_original=FALSE
 
 [Xbstream]
+#EXPERIMENTAL
 # Enable this, if you want to stream your backups
 #xbstream=/usr/bin/xbstream
 #stream=xbstream
 #xbstream_options=-x --parallel=100
+#xbs_decrypt=1
 # WARN, enable this, if you want to stream your backups to remote host
 #remote_stream=ssh xxx.xxx.xxx.xxx
 

--- a/general_conf/bck.conf
+++ b/general_conf/bck.conf
@@ -56,8 +56,9 @@ xtra_prepare=--apply-log-only
 
 [Xbstream]
 # Enable this, if you want to stream your backups
+#xbstream=/usr/bin/xbstream
 #stream=xbstream
-#parallel=16
+#xbstream_options=-x --parallel=100
 # WARN, enable this, if you want to stream your backups to remote host
 #remote_stream=ssh xxx.xxx.xxx.xxx
 

--- a/general_conf/generalops.py
+++ b/general_conf/generalops.py
@@ -101,8 +101,12 @@ class GeneralClass:
                 self.remove_original_enc = ENC['remove_original']
 
             XBS = con['Xbstream']
+            if 'xbstream' in XBS:
+                self.xbstream = XBS['xbstream']
             if 'stream' in XBS:
                 self.stream = XBS['stream']
+            if 'xbstream_options' in XBS:
+                self.xbstream_options = XBS['xbstream_options']
 
 
             CM = con['Commands']

--- a/general_conf/generalops.py
+++ b/general_conf/generalops.py
@@ -107,6 +107,8 @@ class GeneralClass:
                 self.stream = XBS['stream']
             if 'xbstream_options' in XBS:
                 self.xbstream_options = XBS['xbstream_options']
+            if 'xbs_decrypt' in XBS:
+                self.xbs_decrypt = XBS['xbs_decrypt']
 
 
             CM = con['Commands']

--- a/general_conf/generalops.py
+++ b/general_conf/generalops.py
@@ -1,5 +1,3 @@
-#!/opt/Python-3.3.2/bin/python3
-
 import configparser
 from os.path import isfile
 import humanfriendly
@@ -101,7 +99,11 @@ class GeneralClass:
                 self.decrypt = ENC['decrypt']
             if 'remove_original' in ENC:
                 self.remove_original_enc = ENC['remove_original']
-                
+
+            XBS = con['Xbstream']
+            if 'stream' in XBS:
+                self.stream = XBS['stream']
+
 
             CM = con['Commands']
             self.start_mysql = CM['start_mysql_command']

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -537,7 +537,7 @@ class Backup(GeneralClass):
             if hasattr(self, 'stream') and isfile(("%s/%s/inc_backup.stream") % (self.inc_dir, recent_inc)) \
                     and hasattr(self, 'encrypt') \
                     and hasattr(self, 'xbs_decrypt'):
-                logger.debug("Using xbstream to extract and decrypt from full_backup.stream!")
+                logger.debug("Using xbstream to extract and decrypt from inc_backup.stream!")
                 xbstream_command = "%s %s --decrypt=%s --encrypt-key=%s --encrypt-threads=%s " \
                                    "< %s/%s/inc_backup.stream -C %s/%s" % (
                                        self.xbstream,

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -369,14 +369,12 @@ class Backup(GeneralClass):
                                        and hasattr(self, 'xbs_decrypt'):
                 logger.debug("Using xbstream to extract and decrypt from full_backup.stream!")
                 xbstream_command = "%s %s --decrypt=%s --encrypt-key=%s --encrypt-threads=%s " \
-                                   " --encrypt-chunk-size=%s" \
                                    "< %s/%s/full_backup.stream -C %s/%s" % (
                                     self.xbstream,
                                     self.xbstream_options,
                                     self.decrypt,
                                     self.encrypt_key,
                                     self.encrypt_threads,
-                                    self.encrypt_chunk_size,
                                     self.full_dir,
                                     recent_bck,
                                     self.full_dir,

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -17,7 +17,7 @@ from mysql.connector import errorcode
 from sys import exit
 from general_conf.generalops import GeneralClass
 from general_conf.check_env import CheckEnv
-from os.path import join
+from os.path import join, isfile
 from os import makedirs
 
 import logging
@@ -268,6 +268,7 @@ class Backup(GeneralClass):
             args += " "
             args += '--stream="%s"' % self.stream
             args += " > %s/full_backup.stream" % full_backup_dir
+            logger.warning("Streaming is enabled!")
 
         logger.debug("The following backup command will be executed %s", args)
 
@@ -362,6 +363,32 @@ class Backup(GeneralClass):
             if hasattr(self, 'encrypt_chunk_size'):
                 args += " --encrypt-chunk-size=%s" % (self.encrypt_chunk_size)
 
+            # Extracting streamed full backup prior to executing incremental backup
+
+            if hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") % (self.full_dir, recent_bck)):
+                logger.debug("Using xbstream to extract from full_backup.stream!")
+                xbstream_command = "%s %s < %s/%s/full_backup.stream -C %s/%s" % (
+                    self.xbstream,
+                    self.xbstream_options,
+                    self.full_dir,
+                    recent_bck,
+                    self.full_dir,
+                    recent_bck
+                )
+
+                logger.debug(
+                    "The following xbstream command will be executed %s",
+                xbstream_command)
+
+                status, output = subprocess.getstatusoutput(xbstream_command)
+                if status == 0:
+                    logger.debug("XBSTREAM command succeeded.")
+                else:
+                    logger.error("XBSTREAM COMMAND FAILED!")
+                    time.sleep(5)
+                    logger.error(output)
+                    return False
+
             if 'encrypt' in args:
                 logger.debug("Applying workaround for LP #1444255")
                 xbcrypt_command = "%s -d -k %s -a %s -i %s/%s/xtrabackup_checkpoints.xbcrypt " \
@@ -395,6 +422,13 @@ class Backup(GeneralClass):
                args += " "
                args += '--databases="%s"' % (self.partial_list)
                logger.warning("Partial Backup is enabled!")
+
+            # Checking if streaming enabled for backups
+            if hasattr(self, 'stream'):
+               args += " "
+               args += '--stream="%s"' % self.stream
+               args += " > %s/inc_backup.stream" % inc_backup_dir
+               logger.warning("Streaming is enabled!")
 
             logger.debug(
                 "The following backup command will be executed %s", args)

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -533,10 +533,41 @@ class Backup(GeneralClass):
             if hasattr(self, 'encrypt_chunk_size'):
                 args += " --encrypt-chunk-size=%s" % (self.encrypt_chunk_size)
 
+            # Extract and decrypt streamed full backup prior to executing incremental backup
+            if hasattr(self, 'stream') and isfile(("%s/%s/inc_backup.stream") % (self.inc_dir, recent_inc)) \
+                    and hasattr(self, 'encrypt') \
+                    and hasattr(self, 'xbs_decrypt'):
+                logger.debug("Using xbstream to extract and decrypt from full_backup.stream!")
+                xbstream_command = "%s %s --decrypt=%s --encrypt-key=%s --encrypt-threads=%s " \
+                                   "< %s/%s/inc_backup.stream -C %s/%s" % (
+                                       self.xbstream,
+                                       self.xbstream_options,
+                                       self.decrypt,
+                                       self.encrypt_key,
+                                       self.encrypt_threads,
+                                       self.inc_dir,
+                                       recent_bck,
+                                       self.inc_dir,
+                                       recent_inc
+                                   )
+
+                logger.debug(
+                    "The following xbstream command will be executed %s",
+                    xbstream_command)
+
+                status, output = subprocess.getstatusoutput(xbstream_command)
+                if status == 0:
+                    logger.debug("XBSTREAM command succeeded.")
+                else:
+                    logger.error("XBSTREAM COMMAND FAILED!")
+                    time.sleep(5)
+                    logger.error(output)
+                    return False
+
             # Extracting streamed incremental backup prior to executing new incremental backup
 
-            if hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") % (self.full_dir, recent_bck)):
-                logger.debug("Using xbstream to extract from full_backup.stream!")
+            elif hasattr(self, 'stream') and isfile(("%s/%s/inc_backup.stream") % (self.full_dir, recent_bck)):
+                logger.debug("Using xbstream to extract from inc_backup.stream!")
                 xbstream_command = "%s %s < %s/%s/inc_backup.stream -C %s/%s" % (
                     self.xbstream,
                     self.xbstream_options,
@@ -559,7 +590,7 @@ class Backup(GeneralClass):
                     logger.error(output)
                     return False
 
-            if 'encrypt' in args:
+            elif 'encrypt' in args:
                 logger.debug("Applying workaround for LP #1444255")
                 xbcrypt_command = "%s -d -k %s -a %s -i %s/%s/xtrabackup_checkpoints.xbcrypt " \
                                   "-o %s/%s/xtrabackup_checkpoints" % \

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -546,7 +546,7 @@ class Backup(GeneralClass):
                                        self.encrypt_key,
                                        self.encrypt_threads,
                                        self.inc_dir,
-                                       recent_bck,
+                                       recent_inc,
                                        self.inc_dir,
                                        recent_inc
                                    )

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -375,7 +375,7 @@ class Backup(GeneralClass):
                                     self.xbstream_options,
                                     self.decrypt,
                                     self.encrypt_key,
-                                    self.encrypt_key,
+                                    self.encrypt_threads,
                                     self.encrypt_chunk_size,
                                     self.full_dir,
                                     recent_bck,

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -503,6 +503,32 @@ class Backup(GeneralClass):
             if hasattr(self, 'encrypt_chunk_size'):
                 args += " --encrypt-chunk-size=%s" % (self.encrypt_chunk_size)
 
+            # Extracting streamed incremental backup prior to executing new incremental backup
+
+            if hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") % (self.full_dir, recent_bck)):
+                logger.debug("Using xbstream to extract from full_backup.stream!")
+                xbstream_command = "%s %s < %s/%s/inc_backup.stream -C %s/%s" % (
+                    self.xbstream,
+                    self.xbstream_options,
+                    self.inc_dir,
+                    recent_inc,
+                    self.inc_dir,
+                    recent_inc
+                )
+
+                logger.debug(
+                    "The following xbstream command will be executed %s",
+                    xbstream_command)
+
+                status, output = subprocess.getstatusoutput(xbstream_command)
+                if status == 0:
+                    logger.debug("XBSTREAM command succeeded.")
+                else:
+                    logger.error("XBSTREAM COMMAND FAILED!")
+                    time.sleep(5)
+                    logger.error(output)
+                    return False
+
             if 'encrypt' in args:
                 logger.debug("Applying workaround for LP #1444255")
                 xbcrypt_command = "%s -d -k %s -a %s -i %s/%s/xtrabackup_checkpoints.xbcrypt " \
@@ -536,6 +562,13 @@ class Backup(GeneralClass):
                args += " "
                args += '--databases="%s"' % (self.partial_list)
                logger.warning("Partial Backup is enabled!")
+
+           # Checking if streaming enabled for backups
+            if hasattr(self, 'stream'):
+               args += " "
+               args += '--stream="%s"' % self.stream
+               args += " > %s/inc_backup.stream" % inc_backup_dir
+               logger.warning("Streaming is enabled!")
 
             logger.debug(
                 "The following backup command will be executed %s", args)

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -95,7 +95,7 @@ class Backup(GeneralClass):
         command_execute = ' -e "flush logs"'
 
         # Open connection
-        
+
 
         if hasattr(self, 'mysql_socket'):
             command_connection += ' --socket={}'
@@ -363,9 +363,41 @@ class Backup(GeneralClass):
             if hasattr(self, 'encrypt_chunk_size'):
                 args += " --encrypt-chunk-size=%s" % (self.encrypt_chunk_size)
 
-            # Extracting streamed full backup prior to executing incremental backup
+            # Extract and decrypt streamed full backup prior to executing incremental backup
+            if hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") % (self.full_dir, recent_bck))  \
+                                       and hasattr(self, 'encrypt') \
+                                       and hasattr(self, 'xbs_decrypt'):
+                logger.debug("Using xbstream to extract and decrypt from full_backup.stream!")
+                xbstream_command = "%s %s --decrypt=%s --encrypt-key=%s --encrypt-threads=%s " \
+                                   " --encrypt-chunk-size=%s" \
+                                   "< %s/%s/full_backup.stream -C %s/%s" % (
+                                    self.xbstream,
+                                    self.xbstream_options,
+                                    self.decrypt,
+                                    self.encrypt_key,
+                                    self.encrypt_key,
+                                    self.encrypt_chunk_size,
+                                    self.full_dir,
+                                    recent_bck,
+                                    self.full_dir,
+                                    recent_bck
+                                )
 
-            if hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") % (self.full_dir, recent_bck)):
+                logger.debug(
+                    "The following xbstream command will be executed %s",
+                    xbstream_command)
+
+                status, output = subprocess.getstatusoutput(xbstream_command)
+                if status == 0:
+                    logger.debug("XBSTREAM command succeeded.")
+                else:
+                    logger.error("XBSTREAM COMMAND FAILED!")
+                    time.sleep(5)
+                    logger.error(output)
+                    return False
+
+            # Extract streamed full backup prior to executing incremental backup
+            elif hasattr(self, 'stream') and isfile(("%s/%s/full_backup.stream") % (self.full_dir, recent_bck)):
                 logger.debug("Using xbstream to extract from full_backup.stream!")
                 xbstream_command = "%s %s < %s/%s/full_backup.stream -C %s/%s" % (
                     self.xbstream,
@@ -389,7 +421,7 @@ class Backup(GeneralClass):
                     logger.error(output)
                     return False
 
-            if 'encrypt' in args:
+            elif 'encrypt' in args:
                 logger.debug("Applying workaround for LP #1444255")
                 xbcrypt_command = "%s -d -k %s -a %s -i %s/%s/xtrabackup_checkpoints.xbcrypt " \
                                   "-o %s/%s/xtrabackup_checkpoints" % \

--- a/master_backup_script/backuper.py
+++ b/master_backup_script/backuper.py
@@ -260,8 +260,14 @@ class Backup(GeneralClass):
         # Checking if partial recovery list is available
         if hasattr(self, 'partial_list'):
             args += " "
-            args += '--databases="%s"' % (self.partial_list)
+            args += '--databases="%s"' % self.partial_list
             logger.warning("Partial Backup is enabled!")
+
+        # Checking if streaming enabled for backups
+        if hasattr(self, 'stream'):
+            args += " "
+            args += '--stream="%s"' % self.stream
+            args += " > %s/full_backup.stream" % full_backup_dir
 
         logger.debug("The following backup command will be executed %s", args)
 

